frontends/web/src/routes/device/bitbox02/passphrase.tsx

```diff
@@ -35,345 +36,360 @@ const INFO_STEPS_DISABLE = 0;
 
 const CONTENT_MIN_HEIGHT = '38em';
 
-interface MnemonicPassphraseButtonProps {
+type TProps = {
     deviceID: string;
 }
 
-interface State {
-    infoStep: number;
-    passphraseEnabled?: boolean;
-    status: 'info' | 'progress' | 'success';
-    understood: boolean;
-}
+type TStatus = 'info' | 'progress' | 'success';
+type TPassphraseEnabled = boolean | undefined;
 
-type Props = MnemonicPassphraseButtonProps & TranslateProps;
+export const Passphrase = ({ deviceID }: TProps) => {
+  const { t } = useTranslation();
+  const navigate = useNavigate();
+  const [infoStep, setInfoStep] = useState(0);
+  const [status, setStatus] = useState<TStatus>('info');
+  // Can this be a useRef() ?
+  const [passphraseEnabled, setPassphraseEnabled] = useState<TPassphraseEnabled>();
+  const [understood, setUnderstood] = useState(false);
 
-class Passphrase extends Component<Props, State> {
-  public readonly state: State = {
-    infoStep: 0,
-    status: 'info',
-    understood: false,
-  };
-
-  public componentDidMount() {
-    getDeviceInfo(this.props.deviceID)
-      .then(result => {
+  // Sync. component state with device info.
+  useEffect(() => {
+    let ignoreResult = false;
+    getDeviceInfo(deviceID).then(result => {
+      if (!ignoreResult) {
         if (!result.success) {
           console.error(result.message);
-          alertUser(this.props.t('genericError'));
+          alertUser(t('genericError'));
           return;
         }
         const { mnemonicPassphraseEnabled } = result.deviceInfo;
-        this.setState({
-          // before enabling/disabling we show 1 or more pages to inform about the feature
-          // each page has a continue button that jumps to the next or finally toggles passphrase
-          // infoStep counts down in decreasing order
-          infoStep: mnemonicPassphraseEnabled
-            ? INFO_STEPS_DISABLE
-            : INFO_STEPS_ENABLE,
-          passphraseEnabled: mnemonicPassphraseEnabled,
-        });
-      })
-      .catch(console.error);
-  }
+        setPassphraseEnabled(mnemonicPassphraseEnabled);
+        setInfoStep(mnemonicPassphraseEnabled ? INFO_STEPS_DISABLE : INFO_STEPS_ENABLE);
+      }
+    }).catch(console.error);
 
-  private togglePassphrase = async () => {
-    const { deviceID, t } = this.props;
-    const enable = !this.state.passphraseEnabled;
-    this.setState({ status: 'progress' });
+    return () => {
+      // In case the effect runns multiple times, we only conisder the last result,
+      // and ignore the prior ones.
+      ignoreResult = true;
```

Comment:
It seemed like it runs two times in the beginning but I learned that was only because react does this on purpose to force you to have a good cleanup function. in production it should only run once.

also it would re-run if any of the dependencies change `deviceID` `setPassphraseEnabled` `t` but actually the cleanup works well, there is no problem if it runs multiple times (also the 'result' of the api call is always the same anyways.)

I think it is _technically correct_ to have this here but it really does not make much sense. I'll remove it.

Reply:
if you expect that the component can be unmounted but there might still be an api result that is incoming use hooks/mount. But please do this later, this change should only be about class->functional.